### PR TITLE
Implements Google OAuth

### PR DESCRIPTION
**Pull request by hsmitha26**
This PR does the following:
- User can sign-up/login to our app with their Google account credentials.
- Feature test is passing. All tests are passing.
- Uses OmniAuth and Devise provided Omniauthable module to implement OAuth.

Feature is working in development.

### CODE REVIEW

File: app/controllers/users/omniauth_callbacks_controller.rb [hunk 1]
```diff
@@ -4,6 +4,16 @@ class Users::OmniauthCallbacksController < Devise::OmniauthCallbacksController
...
   # devise :omniauthable, omniauth_providers: [:twitter]
 
+  def google_oauth2
+    @user = User.from_omniauth(request.env["omniauth.auth"])
+    if @user.persisted?
+      sign_in @user
+      redirect_to user_path(@user.id), notice: 'Successfully logged in using your Google account.'
+    else
+      session["devise.google_data"] = request.env["omniauth.auth"]
+      redirect_to root_path, alert: 'Something went wrong. Please try to register instead.'
```
Comment (leiyakenney) [inline app/controllers/users/omniauth_callbacks_controller.rb:14]: Can we go over what devise.google_data is doing?

**Review by leiyakenney — APPROVED** [review]
Looks good! Do we know if this factors in to the unique email restriction we have? Or if it at leastvdoesn't allow a person to have two accounts for that one gmail; one through oauth and one through normal sign in?

File: app/controllers/users/omniauth_callbacks_controller.rb (hunk 1 above)
Comment (hsmitha26) [reply app/controllers/users/omniauth_callbacks_controller.rb:14]: I finally found the resource I used to implement this controller. Here's the explanation regarding what session["devise.google_data"] does:
```In case the user is not persisted, we store the OmniAuth data in the session. Notice we store this data using "devise." as key namespace. This is useful because Devise removes all the data starting with "devise." from the session whenever a user signs in, so we get automatic session clean up. At the end, we redirect the user back to our registration form.```
Also, here's the link with more info for future reference: https://github.com/plataformatec/devise/wiki/OmniAuth:-Overview

**Comment by hsmitha26** [comment]
I did not change the uniqueness to :email (from :user_name) because of validation issues. I prioritized implementing OAuth instead.
I will need to add edge case tests to check if we are allowing two unique user names but with the same email addresses.

File: app/controllers/users/omniauth_callbacks_controller.rb (hunk 1 above)
Comment (nathangthomas) [reply app/controllers/users/omniauth_callbacks_controller.rb:14]: Strong work! Good job working with wha we have to implement OAuth. Interesting to see how this is done in tandem with devise.

**Review by nathangthomas — APPROVED** [review]
